File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master {

int preparationTime(std::vector<std::string> const & layers, int prep_time) {
    return layers.size() * prep_time;
}

amount quantities(std::vector<std::string>const & layers) {
    int const grams_per_noodle_layer{50};
    double const liters_per_sauce_layer{0.2};
    amount total_quantity {};
    for (auto const & layer : layers) {
        if (layer == "noodles")
            total_quantity.noodles += grams_per_noodle_layer;
        else if (layer == "sauce")
            total_quantity.sauce += liters_per_sauce_layer;
    }
    return total_quantity;
}

void addSecretIngredient(std::vector<std::string>& myList, std::vector<std::string> const & otherList) {
    addSecretIngredient(myList, otherList.back());
}

std::vector<double> scaleRecipe(std::vector<double> const & original_quantities, int desired_portions) {
    double const original_scale{2.0};
    double const new_scale = desired_portions / original_scale;
    std::vector<double> new_quantities{};
    for (auto const & quantity : original_quantities)
        new_quantities.emplace_back(quantity * new_scale);
    return new_quantities;
}

void addSecretIngredient(std::vector<std::string>& myList, std::string ingredient) {
    myList.back() = ingredient;
}
    
}  // namespace lasagna_master

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

func PreparationTime(layers []string, timePerLayer int) int {
    if timePerLayer == 0 {
        timePerLayer = 2
    }
    return len(layers) * timePerLayer
}

func Quantities(layers []string) (noodles int, sauce float64) {
    const gramsOfNoodles = 50
    const litersOfSauce = 0.2

    for _, layer := range layers {
        if layer == "noodles" {
            noodles += gramsOfNoodles
        } else if layer == "sauce" {
            sauce += litersOfSauce
        }
    }
    return
}

func AddSecretIngredient(friendsList []string, myList []string) {
    friendsLast := len(friendsList) - 1
    myLast := len(myList) - 1

    myList[myLast] = friendsList[friendsLast]
}

func ScaleRecipe(amountsForTwo []float64, portions int) (scaledAmounts []float64) {
    scaledAmounts = make([]float64, len(amountsForTwo))
    for i, amount := range amountsForTwo {
        scaledAmounts[i] = amount * float64(portions) / 2
    }
    return scaledAmounts
}
